src/migrations.rs

```diff
@@ -38,9 +48,117 @@ impl Contract {
 // }
 
 #[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
-pub struct OldContract {
+pub struct OldContractV2 {
     pub posts: Vector<VersionedPost>,
     pub post_to_parent: LookupMap<PostId, PostId>,
     pub post_to_children: LookupMap<PostId, Vec<PostId>>,
     pub label_to_posts: UnorderedMap<String, HashSet<PostId>>,
+    pub access_control: AccessControl,
+}
+
+// From OldContractV2 to Contract
+#[near_bindgen]
+impl Contract {
+    fn unsafe_add_post_authors() {
+        let OldContractV2 {
+            posts,
+            post_to_parent,
+            post_to_children,
+            label_to_posts,
+            access_control,
+        } = env::state_read().unwrap();
+        let authors = UnorderedMap::new(StorageKey::AuthorToAuthorPosts);
+
+        env::state_write(&Contract {
+            posts,
+            post_to_parent,
+            post_to_children,
+            label_to_posts,
+            access_control,
+            authors,
+        });
+    }
+
+    fn unsafe_insert_old_post_authors(start: u64, end: u64) -> StateVersion {
+        let mut contract: Contract = env::state_read().unwrap();
+        let total = contract.posts.len();
+        let end = min(total, end);
+        for i in start..end {
+            let versioned_post = contract.posts.get(i);
+            if let Some(versioned_post) = versioned_post {
+                let post: Post = versioned_post.into();
+                let mut author_posts =
+                    contract.authors.get(&post.author_id).unwrap_or_else(|| HashSet::new());
+                author_posts.insert(post.id);
+                contract.authors.insert(&post.author_id, &author_posts);
+            }
+        }
+        env::state_write(&contract);
+        StateVersion::V3 { done: end == total, migrated_count: end }
+    }
+}
+
+#[derive(BorshDeserialize, BorshSerialize, Debug)]
+enum StateVersion {
+    V1,
+    V2,
+    V3 { done: bool, migrated_count: u64 },
+}
+
+const VERSION_KEY: &[u8] = b"VERSION";
+
+fn state_version_read() -> StateVersion {
+    env::storage_read(VERSION_KEY)
+        .map(|data| {
+            StateVersion::try_from_slice(&data).expect("Cannot deserialize the contract state.")
+        })
+        .unwrap_or(StateVersion::V2) // StateVersion is introduced in production contract with V2 State.
+}
+
+fn state_version_write(version: &StateVersion) {
+    let data = version.try_to_vec().expect("Cannot serialize the contract state.");
+    env::storage_write(VERSION_KEY, &data);
+    near_sdk::log!("Migrated to version: {:?}", version);
+}
+
+#[near_bindgen]
+impl Contract {
+    pub fn unsafe_self_upgrade() {
+        near_sdk::assert_self();
+
+        let contract = env::input().expect("No contract code is attached in input");
+        Promise::new(env::current_account_id()).deploy_contract(contract);
+    }
+
+    pub fn unsafe_migrate() {
```

Comment:
**Bikesheding, feel free to ignore it.** I know we want people to "ignore" these methods, but "unsafe" triggers me. Should we just name these methods like `contract_self_deploy` and `contract_self_state_migrate`? I believe "contract_self" will already make these methods "boring enough".

Reply:
I think `unsafe_*` was mostly to emphasize this method doesn't use `near_bindgen`'s state load deserialization / state serialization and write, parse argument from `env::input`value return, etc. This behavior does not "safely" mutate the state with a near-sdk-rs APIs and implementer should be careful to ensure state consistency